Resteasy Reactive: A class annotated with `@JsonView` breaks compilation

### Describe the bug

Since Quarkus 3.3, a resteasy reactive class annotated with `@JsonView` breaks compilation. An ``IllegalArgumentException` is thrown.

```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.3.1:build (default) on project code-with-quarkus: Failed to build
 quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor#handleJsonAn
notations threw an exception: java.lang.IllegalArgumentException: Not a method
[ERROR]         at org.jboss.jandex.ClassInfo.asMethod(ClassInfo.java:1159)
[ERROR]         at io.quarkus.resteasy.reactive.jackson.deployment.processor.ResteasyReactiveJacksonProcessor.handleJsonAnnotations(Resteas
yReactiveJacksonProcessor.java:234)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[ERROR]         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:858)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

### How to Reproduce?

[reproducer.zip](https://github.com/quarkusio/quarkus/files/12482550/reproducer.zip)

### Output of `java -version`

Java 17

### Quarkus version or git rev

3.3.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/ResteasyReactiveJacksonProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/ResteasyReactiveJacksonProcessor.java
@@ -237,7 +237,7 @@ public class ResteasyReactiveJacksonProcessor {
                     if ((jsonViews == null) || (jsonViews.length == 0)) {
                         continue;
                     }
-                    recorder.recordJsonView(getMethodId(instance.target().asMethod()), jsonViews[0].name().toString());
+                    recorder.recordJsonView(getTargetId(instance.target()), jsonViews[0].name().toString());
                 }
             }
             if (resourceClass.annotationsMap().containsKey(CUSTOM_SERIALIZATION)) {

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/JsonViewOnClassTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/JsonViewOnClassTest.java
@@ -1,0 +1,81 @@
+package io.quarkus.resteasy.reactive.jackson.deployment.test;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.not;
+
+import java.util.function.Supplier;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.fasterxml.jackson.annotation.JsonView;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+class JsonViewOnClassTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(User.class, Views.class, Public.class, Private.class, Mixed.class);
+                }
+            });
+
+    @Test
+    void test() {
+        given().accept("application/json").get("public")
+                .then()
+                .statusCode(200)
+                .body(not(containsString("1")), containsString("test"));
+
+        given().accept("application/json").get("mixed")
+                .then()
+                .statusCode(200)
+                .body(containsString("1"), containsString("test"));
+
+        given().accept("application/json").get("private")
+                .then()
+                .statusCode(200)
+                .body(containsString("1"), containsString("test"));
+    }
+
+    @JsonView(Views.Public.class)
+    @Path("public")
+    public static class Public {
+
+        @GET
+        public User get() {
+            return User.testUser();
+        }
+    }
+
+    @JsonView(Views.Private.class)
+    @Path("private")
+    public static class Private {
+
+        @GET
+        public User get() {
+            return User.testUser();
+        }
+    }
+
+    @JsonView(Views.Public.class)
+    @Path("mixed")
+    public static class Mixed {
+
+        @GET
+        @JsonView(Views.Private.class)
+        public User get() {
+            return User.testUser();
+        }
+    }
+}

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonResource.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonResource.java
@@ -255,35 +255,28 @@ public class SimpleJsonResource extends SuperClass<Person> {
     @GET
     @Path("/user-without-view")
     public User userWithoutView() {
-        return testUser();
+        return User.testUser();
     }
 
     @JsonView(Views.Public.class)
     @GET
     @Path("/user-with-public-view")
     public User userWithPublicView() {
-        return testUser();
+        return User.testUser();
     }
 
     @JsonView(Views.Private.class)
     @GET
     @Path("/user-with-private-view")
     public User userWithPrivateView() {
-        return testUser();
+        return User.testUser();
     }
 
     @CustomSerialization(UnquotedFieldsPersonSerialization.class)
     @GET
     @Path("/invalid-use-of-custom-serializer")
     public User invalidUseOfCustomSerializer() {
-        return testUser();
-    }
-
-    private User testUser() {
-        User user = new User();
-        user.id = 1;
-        user.name = "test";
-        return user;
+        return User.testUser();
     }
 
     @GET

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/User.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/User.java
@@ -9,4 +9,11 @@ public class User {
 
     @JsonView(Views.Public.class)
     public String name;
+
+    public static User testUser() {
+        User user = new User();
+        user.id = 1;
+        user.name = "test";
+        return user;
+    }
 }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/ResteasyReactiveServerJacksonRecorder.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/ResteasyReactiveServerJacksonRecorder.java
@@ -19,8 +19,8 @@ public class ResteasyReactiveServerJacksonRecorder {
     private static final Map<String, Class<?>> customSerializationMap = new HashMap<>();
     private static final Map<String, Class<?>> customDeserializationMap = new HashMap<>();
 
-    public void recordJsonView(String methodId, String className) {
-        jsonViewMap.put(methodId, loadClass(className));
+    public void recordJsonView(String targetId, String className) {
+        jsonViewMap.put(targetId, loadClass(className));
     }
 
     public void recordCustomSerialization(String target, String className) {
@@ -40,6 +40,10 @@ public class ResteasyReactiveServerJacksonRecorder {
                 customDeserializationMap.clear();
             }
         });
+    }
+
+    public static Class<?> jsonViewForClass(Class<?> clazz) {
+        return jsonViewMap.get(clazz.getName());
     }
 
     public static Class<?> jsonViewForMethod(String methodId) {

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/FullyFeaturedServerJacksonMessageBodyReader.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/FullyFeaturedServerJacksonMessageBodyReader.java
@@ -175,6 +175,13 @@ public class FullyFeaturedServerJacksonMessageBodyReader extends JacksonBasicMes
                 Class<?> jsonViewValue = ResteasyReactiveServerJacksonRecorder.jsonViewForMethod(resourceInfo.getMethodId());
                 if (jsonViewValue != null) {
                     return effectiveReader.withView(jsonViewValue);
+                } else {
+                    jsonViewValue = ResteasyReactiveServerJacksonRecorder
+                            .jsonViewForClass(resourceInfo.getResourceClass());
+                    if (jsonViewValue != null) {
+                        return effectiveReader.withView(jsonViewValue);
+                    }
+
                 }
             }
         }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/FullyFeaturedServerJacksonMessageBodyWriter.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/FullyFeaturedServerJacksonMessageBodyWriter.java
@@ -66,6 +66,13 @@ public class FullyFeaturedServerJacksonMessageBodyWriter extends ServerMessageBo
                 Class<?> jsonViewValue = ResteasyReactiveServerJacksonRecorder.jsonViewForMethod(resourceInfo.getMethodId());
                 if (jsonViewValue != null) {
                     effectiveWriter = effectiveWriter.withView(jsonViewValue);
+                } else {
+                    jsonViewValue = ResteasyReactiveServerJacksonRecorder
+                            .jsonViewForClass(resourceInfo.getResourceClass());
+                    if (jsonViewValue != null) {
+                        effectiveWriter = effectiveWriter.withView(jsonViewValue);
+                    }
+
                 }
             }
             effectiveWriter.writeValue(stream, o);

